doc/man/gnunet-setup.1
.TH GNUNET-SETUP "1" "8 Dec 2007" "GNUnet"
.SH NAME
gnunet\-setup \- configuring GNUnet
.SH SYNOPSIS
.B gnunet\-setup
[\fIOPTIONS\fR] MODE
.SH DESCRIPTION
.PP
gnunet\-setup can be used to generate and edit the configuration files (gnunetd.conf and gnunet.conf) for GNUnet.  gnunet\-setup runs in two basic modes, one for editing gnunet.conf and one for editing gnunetd.conf.  Depending on which libraries were available when gnunet\-setup was build, gnunet\-setup will have a curses or GTK user interface.  A text\-only interface is always available.  For a quick initial configuration gnunet\-setup offers a wizard (GTK only) that can be used to quickly generate a working gnunetd.conf file.
.TP
\fB\-h\fR, \fB\-\-help\fR
print help page
.TP
\fB\-c \fIFILENAME\fR, \fB\-\-config=FILENAME\fR
specify location of the configuration file (default: /etc/gnunetd.conf or ~/.gnunet/gnunet.conf, depending on the mode)
.TP
\fB\-d\fR, \fB\-\-daemon\fR
generate configuration for gnunetd (gnunetd.conf).  If run without this option, gnunet\-setup will generate a configuration for GNUnet clients.
.TP
\fB\-g \fISECTION:OPTION\fR, \fB\-\-get=\fISECTION:OPTION\fR
display the value of the given OPTION from the given SECTION in the configuration file.
.TP
\fB\-s \fISECTION:OPTION=VALUE\fR, \fB\-\-set=\fISECTION:OPTION=VALUE\fR
set the value of the given OPTION from the given SECTION to the specified VALUE.  If this change implies other options also changing, perform those changes as well.
.TP
\fB\-v\fR, \fB\-\-version\fR
print the version number
.TP
.SH NOTES
Before you can use any GNUnet application you must run gnunet\-setup to generate the configuration.  First, run gnunet\-setup with the \-d option to generate a configuration for gnunetd.  Later, each user of the system should run gnunet\-setup without \-d to generate his personal configuration.
.TP
Depending on your system, gnunet\-setup will have a subset of the following modes: config (text), menuconfig (curses), gconfig (GTK), and wizard\-gtk (GTK wizard).  The wizards can currently only be used in conjunction with the \-d option to configure gnunetd.
.SH FILES
.TP
/etc/gnunetd.conf
gnunetd configuration file (default location).
.TP
~/.gnunet/gnunet.conf
client configuration file (default location).
.SH "REPORTING BUGS"
Report bugs by using mantis <https://gnunet.org/bugs/> or by sending electronic mail to <gnunet-developers@gnu.org>
.SH "SEE ALSO"
\fBgnunetd\fP(1), \fBgnunet\-update\fP(1), \fBgnunetd.conf\fP(5), \fBgnunet.conf\fP(5)


ChangeLog
Mon Nov 16 09:25:47 CET 2009
	Allow passing of command-line arguments using GNUNET_ARGS
	environment variable (see Mantis #1467).

Mon Apr 27 22:18:19 CEST 2009
	Monitor low-traffic connections over a longer period
	of time before terminating them.

Fri Feb 27 22:05:50 MST 2009
	Releasing 0.8.0c.

Sun Feb  1 19:51:40 MST 2009
	Fixed Mantis #1429 (struct padding could cause sizeof
	to be different on AMD64, using gcc-ism "packed" to
	avoid this problem).

Sun Feb  1 00:37:16 MST 2009
	Fixed problem in DHT client protocol that could cause
	PUT requests from clients to be (frequently) lost.

Sun Nov  2 16:06:47 MST 2008
	Fixed problem with MySQL disconnects in one module
	impacting MySQL correctness in another module (caused
	crashes).

Sat Aug 16 00:18:34 CEST 2008
	Improved keyword extraction
	Fixed expiration time of keywords
	Releasing 0.8.0b.

Sun Jul 20 15:27:11 MDT 2008
	Releasing 0.8.0a.

Thu Jul 10 00:59:31 MDT 2008
	Fixed bug where datastore resizing would not 
	result in bloomfilter update during gnunet-update.

Sat Jun 28 23:43:00 MDT 2008
	Fixed bug preventing namespace advertisements from
	being processed.

Sat Jun 14 00:54:39 MDT 2008
	Releasing 0.8.0.

Fri Jun 13 23:51:00 MDT 2008
	GNUnet now compiles on OpenBSD.

Wed Jun 11 22:17:09 MDT 2008
	Made state of gnunet-auto-share persistent. 
	Added support for multiple directories.  Improved
	FSUI/ECRS APIs to reduce the number of threads needed
	for probing.  Increased number of concurrent probes
	allowed.

Sat Jun  7 17:54:49 MDT 2008
	Simplified namespaces a lot.  Shorter timeouts for
	trying to connect to gnunetd to prevent interactive
	apps from blocking too long when gnunetd is not
	running.

Sun Jun  1 21:11:54 MDT 2008
	Releasing 0.8.0pre1.

Sun Jun  1 13:41:27 MDT 2008
	Split of libgnunetpseudonym from libgnunetnamespace
	(with major API extension and changes).

Tue May  6 04:10:02 MDT 2008
	Added simple sanity check for [MODULES] values.
	Added "-s" and "-q" options to gnunet-peer-info.

Thu Apr 24 00:05:17 MDT 2008
	Simplified FSUI stop/abort/pause/resume API calls
	(no need to pass FSUI context anymore).

Mon Apr 21 21:05:20 MDT 2008
	Adding buffered IO for FSUI's serialize and
	deserialize code.

Sun Apr 20 20:01:20 MDT 2008
	Fixed bug in HTTP transport causing lonely messages
	(responses to GET) not to be delivered.

Mon Mar 24 21:15:36 MDT 2008
	Fixed various problems with downloading locally
	indexed large files (downloading large files from
	other peers should not have been a problem).
	
Sun Mar  2 15:33:41 MST 2008
	Added option to allow user to disable IPv6 support.
	Releasing 0.8.0pre0.

Sat Mar  1 21:14:02 MST 2008
	Fixed some dstore performance issues.
	Made transports alternate between trying IPv4 and IPv6
	if both are available (since one of them maybe
	misconfigured).

Mon Feb 25 23:25:48 MST 2008
	Cleanup of the DNS code: removed GNUNET_IPvXAddress
	and GNUNET_PRIP; centralized all DNS lookup code in 
	util/network/dns.c; improved IPv6 support for 
        network-client code.

Mon Feb 25 00:01:27 MST 2008
	Added asynchronous search and download methods for 
	ECRS library.  FSUI now can do with only one thread
	per search or download (until now, we had two 
	threads per search / download).

Tue Feb 19 20:35:28 MST 2008
	Updated database schemata to support O(1) operations
	even if there are N files under the same keywords.
	This breaks database backwards compatibility (and
	no migration code is provided).

Mon Feb 18 19:47:37 MST 2008
	Added new options -u and -s to gnunet-insert.

Sat Feb 16 21:37:33 MST 2008
	Implemented gnunet-auto-share for automatic
	sharing of directories.

Thu Feb 14 01:02:32 MST 2008
	Made HTTP transport work over IPv6 as well.

Wed Feb 13 22:59:07 MST 2008
	Modified mySQL and SQLite datastores to return diverse
	result sets during (partial) hash-based get iterations
	(should result in more diverse keyword result sets).

Tue Feb 12 23:54:34 MST 2008
	First shot at integrated hostlist HTTP server based
	on libmicrohttpd.

Sat Feb  9 22:06:23 MST 2008
	Minor changes to the ECRS/FSUI APIs for searches
	(no more timeouts and/or max-results).  Clients who
	need these features could easily implement them 
	directly (and they are not really useful to begin
	with).  
	Also, the IPv4 and IPv6 protocols are being integrated
	so that we only listen to one port (running both IPv4
	and IPv6 over the same port).  This break transport
	compatibility, but given that we're breaking protocol
	compatibility for file-sharing already, this may 
	actually be a bit of a good thing (since it will now
	be possible to tell if peers running the new GAP code
	are connected to other peers running the new GAP code).
	
Sun Feb  3 13:17:09 MST 2008
	Dramatic changes to the GAP implementation (breaking
	protocol compatibility).  Essentially, we can save
	a few bytes in each reply.  More importantly, the new
	code allows the searching client to specify a set of
	replies that are not desired (hopefully helping to
	dramatically increase the diversity of search replies
	obtained over time).  Note that the actual encoding
	and databases are not affected (just P2P protocol).
	The update is not complete yet, but should compile.

Tue Jan  8 20:07:20 MST 2008
	Added option for testing ("make check") to use weak(er)
	PRNG for key generation (thanks to Werner Koch for 
	pointing out how to do it).

Thu Dec 22 20:10:37 MST 2007
	Releasing GNUnet 0.7.3.

Sun Dec  9 14:34:32 MST 2007
	Implemented MySQL version of the dstore-module.  This
	means that sqLite is now again truly optional.

Sat Dec  8 15:15:53 MST 2007
	Integrated F2F topology into main topology module 
	(options are used to choose between F2F-only (old F2F), 
	 minimum number of friend connections (NEW) or entirely
	 arbitrary connection set (default)).

Thu Dec  6 20:51:17 MST 2007
	The SMTP transport is working again.

Thu Nov 22 20:49:16 MST 2007
	United all libgnunetutil_*.so libraries into one big
	libgnunetutil.so library (eliminates issues for binary

src/setup/gnunet-setup.c
/*
     This file is part of GNUnet.
     (C) 2001, 2002, 2005, 2006, 2007 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file setup/gnunet-setup.c
 * @brief GNUnet Setup
 * @author Nils Durner
 * @author Christian Grothoff
 */
#include "platform.h"
#include "gnunet_util.h"
#include "gnunet_directories.h"
#include "gnunet_setup_lib.h"

typedef int (*ConfigurationPluginMain) (int argc,
                                        char *const *argv,
                                        struct GNUNET_PluginHandle * self,
                                        struct GNUNET_GE_Context * ectx,
                                        struct GNUNET_GC_Configuration * cfg,
                                        struct GNUNET_GNS_Context * gns,
                                        const char *filename, int is_daemon);

static int config_daemon;

static struct GNUNET_GE_Context *ectx;

static struct GNUNET_GC_Configuration *cfg;

#if HAVE_GUILE
static struct GNUNET_GNS_Context *gns;
#endif

static char *cfgFilename;

static int option_processing;

static char *get_option;

static char *set_option;

static int
set_option_helper (GNUNET_CommandLineProcessorContext * ctx,
                   void *unused, const char *cmdLineOption, const char *value)
{
  option_processing = GNUNET_YES;
  if (set_option != NULL)
    {
      fprintf (stderr, _("Can only set one option per invocation.\n"));
      return GNUNET_SYSERR;
    }
  if ((NULL == strstr (value, ":")) ||
      (NULL == strstr (strstr (value, ":"), "=")))
    {
      fprintf (stderr,
               _
               ("Invalid syntax, argument to 'set' must have the format SECTION:OPTION=VALUE.\n"));
      return GNUNET_SYSERR;
    }
  set_option = GNUNET_strdup (value);
  return GNUNET_OK;
}

static int
get_option_helper (GNUNET_CommandLineProcessorContext * ctx,
                   void *unused, const char *cmdLineOption, const char *value)
{
  option_processing = GNUNET_YES;
  if (get_option != NULL)
    {
      fprintf (stderr, _("Can only display one option per invocation.\n"));
      return GNUNET_SYSERR;
    }
  if (NULL == strstr (value, ":"))
    {
      fprintf (stderr,
               _
               ("Invalid syntax, argument to 'get' must have the format SECTION:OPTION.\n"));
      return GNUNET_SYSERR;
    }
  get_option = GNUNET_strdup (value);
  return GNUNET_OK;
}


/**
 * All gnunet-setup command line options
 */
static struct GNUNET_CommandLineOption gnunetsetupOptions[] = {
  GNUNET_COMMAND_LINE_OPTION_CFG_FILE (&cfgFilename),   /* -c */
  {'d', "daemon", NULL,
   gettext_noop ("generate configuration for gnunetd, the GNUnet daemon"),
   0, &GNUNET_getopt_configure_set_one, &config_daemon},
  {'g', "get", "SECTION:ENTRY",
   gettext_noop ("print a value from the configuration file to stdout"),
   1, &get_option_helper, NULL},
  GNUNET_COMMAND_LINE_OPTION_HELP (gettext_noop ("Tool to setup GNUnet.")),     /* -h */
  {'s', "set", "SECTION:ENTRY=VALUE",
   gettext_noop ("update a value in the configuration file"),
   1, &set_option_helper, NULL},
  GNUNET_COMMAND_LINE_OPTION_VERSION (PACKAGE_VERSION), /* -v */
  GNUNET_COMMAND_LINE_OPTION_END,
};

/**
 * Client or server-only gnunet-setup command line options
 */
static struct GNUNET_CommandLineOption gnunetsetupNoDOptions[] = {
  GNUNET_COMMAND_LINE_OPTION_CFG_FILE (&cfgFilename),   /* -c */
  {'g', "get", "SECTION:ENTRY",
   gettext_noop ("print a value from the configuration file to stdout"),
   1, &get_option_helper, NULL},
  GNUNET_COMMAND_LINE_OPTION_HELP (gettext_noop ("Tool to setup GNUnet.")),     /* -h */
  {'s', "set", "SECTION:ENTRY=VALUE",
   gettext_noop ("update a value in the configuration file"),
   1, &set_option_helper, NULL},
  GNUNET_COMMAND_LINE_OPTION_VERSION (PACKAGE_VERSION), /* -v */
  GNUNET_COMMAND_LINE_OPTION_END,
};

#if HAVE_GUILE
static void
gns2cfg (struct GNUNET_GNS_TreeNode *pos)
{
  int i;
  char *val;

  if (pos == NULL)
    return;
  i = 0;
  while (pos->children[i] != NULL)
    {
      gns2cfg (pos->children[i]);
      i++;
    }
  if (i != 0)
    return;
  if ((pos->section == NULL) || (pos->option == NULL))
    return;
  if (GNUNET_NO ==
      GNUNET_GC_have_configuration_value (cfg, pos->section, pos->option))
    {
      val = GNUNET_GNS_get_default_value_as_string (pos->type, &pos->value);
      if (val != NULL)
        {
          GNUNET_GC_set_configuration_value_string (cfg,
                                                    ectx,
                                                    pos->section, pos->option,
                                                    val);
          GNUNET_free (val);
        }
    }
}
#endif

static int
dyn_config (const char *module,
            const char *mainfunc,
            int argc, char *const *argv, const char *filename)
{
  ConfigurationPluginMain mptr;
  struct GNUNET_PluginHandle *library;
  library = GNUNET_plugin_load (ectx, "libgnunet", module);
  if (!library)
    return GNUNET_SYSERR;
  mptr = GNUNET_plugin_resolve_function (library, mainfunc, GNUNET_YES);
  if (!mptr)
    {
      GNUNET_plugin_unload (library);
      return GNUNET_SYSERR;
    }
  mptr (argc, argv, library, ectx, cfg,
#if HAVE_GUILE
        gns,
#else
        NULL,
#endif
        filename, config_daemon);
  GNUNET_plugin_unload (library);
  return GNUNET_YES;
}

static const char *INFO = "gnunet-setup [OPTIONS] config|generate-defaults"
#if HAVE_DIALOG
#if HAVE_GUILE
  "|menuconfig"
#endif
  "|wizard-curses"
#endif
#if HAVE_GTK
#if HAVE_GUILE
  "|gconfig"
#endif
  "|wizard-gtk"
#endif
#if HAVE_QT
  "|wizard-qt"
#endif
#if OSX
  "|config-cocoa"
#endif
  "";

static const char *INFO_CLIENT_ONLY =
  "gnunet-setup [OPTIONS] config"
#if HAVE_GUILE
  "|generate-defaults"
#endif
#if HAVE_DIALOG
  "|menuconfig"
#endif
#if HAVE_GTK
#if HAVE_GUILE
  "|gconfig"
#endif
#endif
  "";

/**
 * List of supported plugins.  One entry consists
 * of three strings: option name, plugin library
 * name and main method name.
 */
static const char *modules_all[] = {
  "gconfig", "setup_gtk", "gconf_main",
  "config-cocoa", "setup_cocoa", "config_cocoa_main",
  "menuconfig", "setup_curses", "mconf_main",
  "config", "setup_text", "main_",
  "wizard-curses", "setup_curses", "wizard_curs_main",
  "wizard-gtk", "setup_gtk", "gtk_wizard_main",
  "wizard-qt", "setup_qt", "qt_wizard_main",
  "generate-defaults", "setup_text", "dump_",
  NULL,
};

/**
 * List of supported plugins if we do not
 * have the server part.  One entry consists
 * of three strings: option name, plugin library
 * name and main method name.
 */
static const char *modules_client_only[] = {
  "gconfig", "setup_gtk", "gconf_main",
  "menuconfig", "setup_curses", "mconf_main",
  "config", "setup_text", "main_",
  "generate-defaults", "setup_text", "dump_",
  NULL,
};


int
main (int argc, char *const *argv)
{
  const char *operation;
  int done;
  char *dirname;
#if HAVE_GUILE
  char *specname;
#endif
  char *value;
  char *option;
  int i;
  struct stat buf;
  int have_files;
  const char **modules;

  ectx = GNUNET_GE_create_context_stderr (GNUNET_NO,
                                          GNUNET_GE_WARNING | GNUNET_GE_ERROR
                                          | GNUNET_GE_FATAL | GNUNET_GE_USER |
                                          GNUNET_GE_ADMIN |
                                          GNUNET_GE_DEVELOPER |
                                          GNUNET_GE_IMMEDIATE |
                                          GNUNET_GE_BULK);
  GNUNET_GE_setDefaultContext (ectx);
  GNUNET_os_init (ectx);

#if ENABLE_NLS
  setlocale (LC_ALL, "");
  dirname = GNUNET_get_installation_path (GNUNET_IPK_LOCALEDIR);
  if (dirname != NULL)
    {
      BINDTEXTDOMAIN ("GNUnet", dirname);
      GNUNET_free (dirname);
    }
  textdomain ("GNUnet");
#endif

  cfg = GNUNET_GC_create ();
  GNUNET_GE_ASSERT (ectx, cfg != NULL);

  have_files = 0;
#if HAVE_GUILE
  dirname = GNUNET_get_installation_path (GNUNET_IPK_DATADIR);
  GNUNET_GE_ASSERT (ectx, dirname != NULL);
  specname =
    GNUNET_malloc (strlen (dirname) + strlen ("config-daemon.scm") + 1);
  strcpy (specname, dirname);
  strcat (specname, "config-daemon.scm");
  if (0 == ACCESS (specname, R_OK))
    have_files = 1;
  strcpy (specname, dirname);
  strcat (specname, "config-client.scm");
  if (0 == ACCESS (specname, R_OK))
    have_files |= 2;
  GNUNET_free (specname);
  GNUNET_free (dirname);
#endif
  if (have_files == 1)
    config_daemon = GNUNET_YES;

  i = GNUNET_parse_options ((have_files == 2) ? INFO_CLIENT_ONLY : INFO,
                            ectx,
                            cfg,
                            (have_files ==
                             3) ? gnunetsetupOptions : gnunetsetupNoDOptions,
                            (unsigned int) argc, argv);
  if (i < 0)
    {
      GNUNET_GC_free (cfg);
      GNUNET_GE_free_context (ectx);
      return -1;
    }
  if (option_processing)
    {
      operation = "options";
    }
  else
    {
      if (i != argc - 1)
        {
          if (i < argc - 1)
            {
              fprintf (stderr, _("Too many arguments.\n"));
              return -1;
            }
          GNUNET_GE_LOG (ectx,
                         GNUNET_GE_WARNING | GNUNET_GE_REQUEST |
                         GNUNET_GE_USER,
                         _("No interface specified, using default.\n"));
          operation = "config";
#if HAVE_DIALOG
          operation = "menuconfig";
#endif
#if OSX
          operation = "config-cocoa";
#endif
#if HAVE_GTK
          operation = "gconfig";
#endif
        }
      else
        {
          operation = argv[i];
        }
    }
  if (NULL != strstr (operation, "wizard"))
    config_daemon = GNUNET_YES; /* wizard implies daemon! */
  if (cfgFilename == NULL)
    cfgFilename = config_daemon
      ? GNUNET_strdup (GNUNET_DEFAULT_DAEMON_CONFIG_FILE)
      : GNUNET_strdup (GNUNET_DEFAULT_CLIENT_CONFIG_FILE);
  dirname = GNUNET_expand_file_name (ectx, cfgFilename);
  GNUNET_free (cfgFilename);
  cfgFilename = GNUNET_strdup (dirname);
  i = strlen (dirname) - 1;
  while (i > -1)
    {
      char ch = dirname[i];
      if ((ch == '/') || (ch == '\\'))
        {
          dirname[i + 1] = 0;
          break;
        }
      i--;
    }
  GNUNET_disk_directory_create (ectx, dirname);
  if (((0 != ACCESS (cfgFilename, W_OK)) &&
       ((errno != ENOENT) || (0 != ACCESS (dirname, W_OK)))))
    GNUNET_GE_DIE_STRERROR_FILE (ectx,
                                 GNUNET_GE_FATAL | GNUNET_GE_USER |
                                 GNUNET_GE_ADMIN | GNUNET_GE_IMMEDIATE,
                                 "access", dirname);
  if ((0 == STAT (cfgFilename, &buf)) && (S_ISDIR (buf.st_mode)))
    {
      fprintf (stderr,
               _
               ("Configuration file `%s' must be a filename (but is a directory).\n"),
               cfgFilename);
      GNUNET_GC_free (cfg);
      GNUNET_GE_free_context (ectx);
      GNUNET_free (dirname);
      return 1;
    }
  GNUNET_free (dirname);

  if (0 == ACCESS (cfgFilename, F_OK))
    if (0 != GNUNET_GC_parse_configuration (cfg, cfgFilename))
      fprintf (stderr,
	       _("Failed to parse configuration file `%s'\n"),
	       cfgFilename);
  dirname = GNUNET_get_installation_path (GNUNET_IPK_DATADIR);
  GNUNET_GE_ASSERT (ectx, dirname != NULL);

#if HAVE_GUILE
  specname =
    GNUNET_malloc (strlen (dirname) + strlen ("config-daemon.scm") + 1);
  strcpy (specname, dirname);
  GNUNET_free (dirname);
  if (config_daemon)
    strcat (specname, "config-daemon.scm");
  else
    strcat (specname, "config-client.scm");
  gns = GNUNET_GNS_load_specification (ectx, cfg, specname);
  GNUNET_free (specname);
  if (gns == NULL)
    {
      GNUNET_GC_free (cfg);
      GNUNET_GE_free_context (ectx);
      GNUNET_free (cfgFilename);
      return -1;
    }
  gns2cfg (GNUNET_GNS_get_tree_root (gns));
#endif

  if (option_processing)
    {
      done = 0;
      if (get_option != NULL)
        {
          option = strstr (get_option, ":");
          option[0] = '\0';
          option++;
          if (GNUNET_NO == GNUNET_GC_have_configuration_value (cfg,
                                                               get_option,
                                                               option))
            {
              fprintf (stderr, _("Undefined option.\n"));
              done = 1;
            }
          else
            {
              GNUNET_GC_get_configuration_value_string (cfg,
                                                        get_option,
                                                        option, NULL, &value);
              fprintf (stdout, "%s\n", value);
              GNUNET_free (value);
            }
          GNUNET_free (get_option);
        }
      if (set_option != NULL)
        {
          option = strstr (set_option, ":");
          option[0] = '\0';
          option++;
          value = strstr (option, "=");
          value[0] = '\0';
          value++;
          if ((GNUNET_OK !=
               GNUNET_GC_set_configuration_value_string (cfg, ectx,
                                                         set_option, option,
                                                         value))
              || (GNUNET_OK !=
                  GNUNET_GC_write_configuration (cfg, cfgFilename)))
            done = 1;
          GNUNET_free (set_option);
        }
#if HAVE_GUILE
      GNUNET_GNS_free_specification (gns);
#endif
      GNUNET_GC_free (cfg);
      GNUNET_GE_free_context (ectx);
      GNUNET_free (cfgFilename);
      return done;
    }
  else
    {
      if (have_files == 2)
        modules = modules_client_only;
      else
        modules = modules_all;
      done = GNUNET_NO;
      i = 0;
      while ((done == GNUNET_NO) && (modules[i] != NULL))
        {
          if (strcmp (operation, modules[i]) == 0)
            {
              if (dyn_config (modules[i + 1],
                              modules[i + 2], argc, argv,
                              cfgFilename) != GNUNET_YES)
                {
                  GNUNET_GE_LOG (ectx,
                                 GNUNET_GE_FATAL | GNUNET_GE_USER |
                                 GNUNET_GE_ADMIN | GNUNET_GE_IMMEDIATE,
                                 _("`%s' is not available.\n"), operation);
#if HAVE_GUILE
                  GNUNET_GNS_free_specification (gns);
#endif
                  GNUNET_GC_free (cfg);
                  GNUNET_GE_free_context (ectx);
                  GNUNET_free (cfgFilename);
                  return -1;
                }
              else
                {
                  done = GNUNET_YES;
                }
            }
          i += 3;
        }
    }
  GNUNET_free (cfgFilename);
  if (done == GNUNET_NO)
    {
      fprintf (stderr, _("Unknown operation '%s'.\n"), operation);
      fprintf (stderr, _("Use --help to get a list of options.\n"));
#if HAVE_GUILE
      GNUNET_GNS_free_specification (gns);
#endif
      GNUNET_GC_free (cfg);
      GNUNET_GE_free_context (ectx);
      return 1;
    }
#if HAVE_GUILE
  GNUNET_GNS_free_specification (gns);
#endif
  GNUNET_GC_free (cfg);
  GNUNET_GE_free_context (ectx);
  return 0;
}
